### PR TITLE
Introduce distraction free mode

### PR DESCRIPTION
**Pull request by draganescu**
## What?

This PR is the result of the exploration in #38928 to introduce a 
distraction free mode. This mode build on top of the reduce interface 
preference, amplifying its effect.

## Why?

To enhance content production in WordPress. Text based content is one of 
the most important forms of content. While the site and page building 
enhancements make multi media content production much easier, through the 
visual tools offered by the block editor, text content production can 
sometimes feel impaired.

A distraction free mode (the term became somewhat of an industry standard) 
aims to remove visual clutter when the focus is on text based content 
production.

## How?

The PR builds on the reduce interface prefference. Until now the 
preference, when toggled on, would only hide some of the toolbar icons and 
a few other non-essential UI items. With this PR the prefference will have 
a more drastic effect, turning the UI into a blank page:

- hides the top toolbar, makes it appear on hover over its normal occupied 
area
- removes many top toolbar buttons
- automatically closes any open sidebars
- hides the insertion point indicator
- hides the block toolbar

## Known issues

- [x] There is a mystery bug that makes the page scroll to top when 
distraction free mode is on and the `ESC` key is pressed 
- [x] There are some failing tests
- [x] Move focus from the options menu to somewhere useful (like the options toggle)
- [x] On small viewports the setting should have no effect, but it does 
break some things
- - The mobile viewport size hides all the "View" group of options in the Editor Options Menu
- [x] The copy is still unsettled (distraction free, toggle interface ...)
- [x] The `hasReducedUi` props need to be refactored into `isDistractionFree`
- [x] Shift+Tab or Alt+F10  should show the toolbar
- [x] The system notifications can't be closed because toolbar is on top
- [x] The chrome should be animated off screen not simply disappear. 
- [x] Snackbar notifications should adjust their bottom position

## Testing Instructions

1. Create a new post
2. Go to top bar > tools menu and click "Toggle interface"
3. Observe serenity

## Screenshots or screencast 

https://user-images.githubusercontent.com/107534/189492067-f2b797fc-a363-47e1-bbac-c96b89d48a4e.mp4

### CODE REVIEW

**Comment by jameskoster** [comment]
This looks like a good start to me. Since it's one part of some other "interface personalisation" ideas, maybe we should open a tracking issue that encapsulates everything? 

In terms of this PR specifically, I don't think that we necessarily need to _disable_ other view options when the interface is hidden. It doesn't seem harmful to allow folks to toggle the top toolbar option for example. It would also be good if the top toolbar remained visible when the ellipsis menu is open.

Edit: It would be good to restore the animations from https://github.com/WordPress/gutenberg/pull/38928#issuecomment-1086600989 as well, and refine. The notion of elements moving in and out of the viewport (rather than simply fading off the canvas) will feel more cohesive when we consider other views like the zoomed out one, and browse mode.

**Comment by draganescu** [comment]
I think it would be good to land this as an infrastructure base to build on. In the exploration PR there was a lot of back and forth to this point and even now the size of the change is considerable. So I wonder what is the minimum lacking here to land this and then start refining?

**Comment by alexstine** [comment]
I'd like to check this out for at least basic A11Y. I'll be around shortly.

**Comment by alexstine** [comment]
@draganescu Could you please fix the hidden toolbars not showing up on keyboard shortcut press? That feels like a bug to me. I understand why you would do this for mouse users as hovering the mouse is easier, but pressing Shift+Tab or Alt+F10 is an intentional action. One that should allow the block toolbar to show up.

Other than that, I am okay with the other changes. Looks good.

Thanks.

**Comment by javierarce** [comment]
I have noticed that the toolbar makes it impossible to close the autosave message:

https://user-images.githubusercontent.com/4933/178979702-1af69fd7-dd15-4790-a485-fd27af35b0af.mov

**Comment by paaljoachim** [comment]
It looks like a very good start!
Thank you Andrei!

One thing. What about adding an X in the top left corner to signal an escape out of this mode?
As a quick way instead of clicking the 3 dots drop down menu and exit out of the distraction free mode.

<details>
<summary>
Mockup
</summary>

Distraction-Free-X

</details>

**Comment by javierarce** [comment]
Here's an idea to solve the issue I mentioned [here](https://github.com/WordPress/gutenberg/pull/41740#issuecomment-1184373886)

https://user-images.githubusercontent.com/4933/187182748-b7e92b92-49d7-4e5a-a894-637bab06c2ec.mov

The video above shows a small non-interactive notification indicating that the user has a pending warning. If the user moves their cursor to the top of the screen the top bar appears, revealing a button that toggles the notification.

If the user moves their cursor outside of the notification, both the notification and the top bar disappear, and we are back to the initial state. 

I'm unsure if there are more complex use cases (for example, with more than one notification at the same time), but please let me know, and I'll incorporate them into the design. 

[Interactive prototype](https://www.figma.com/proto/5PH8mL49tclKKl7z3hRzg0/Gutenberg-Design-Template-(Community)?page-id=1%3A2&node-id=709%3A3466&scaling=min-zoom&starting-point-node-id=709%3A3466&show-proto-sidebar=1).

**Comment by javierarce** [comment]
A small detail that I've noticed testing this PR is that the snackbar message that appears after saving the post has the same bottom margin as in the other modes where there's a breadcrumb. We should have that into account and reduce that bottom margin.

**Comment by jasmussen** [comment]
The yellow notices strike again 😅 — we really should get those updated to become snackbars instead (like Javi's suggestion but perhaps just the default gray style), persistent if need be.

In the mean time, as a very very small interim solution that may be easier to implement, could we just raise the z-index of the notice to be above the top bar? That way you HAVE to dismiss it before you can do anything else?

**Comment by draganescu** [comment]
In https://github.com/WordPress/gutenberg/pull/41740/commits/0f07b543438dcf85a04c50ded748dd35dc23c221 I tried a combination between @javierarce's design and @jasmussen 's suggestion.

---

@javierarce I've fixed the snackbar bottom position. I was wondering if the special behavior for the system notifications is not a bit too complex for this power user mode? Also I think some of these system notifications should not be hidden even in distraction free mode (e.g. the ones about autosaves being more recent). Hence my implementation being on the simple side. Is it bad?

---

> could we just raise the z-index of the notice to be above the top bar? That way you HAVE to dismiss it before you can do anything else?

@jasmussen the problem with this approach is that I had to move the "editor" notifications from the content area of the interface skeleton to the header area - this is likely causing a11y problems already.

---

@alexstine I think it'd be impossible to instruct motion to stop b/c of various areas being focused via Keyboard. I think the best thing I can implement is that for those keyboard shortcuts we exit the distraction free mode? Are there any other situations that should exit this mode?

**Comment by javierarce** [comment]
> @javierarce I've fixed the snackbar bottom position. I was wondering if the special behavior for the system notifications is not a bit too complex for this power user mode? Also I think some of these system notifications should not be hidden even in distraction free mode (e.g. the ones about autosaves being more recent). Hence my implementation being on the simple side. Is it bad?

I think @jasmussen suggestion and your implementation are good… but I see a 25px right margin that would be great to remove.

Screenshot 2022-09-12 at 14 57 42

File: packages/block-editor/src/components/block-tools/insertion-point.js [hunk 1]
```diff
@@ -60,13 +64,17 @@ function InsertionPointPopover( {
...
+
 		return {
 			previousClientId: _previousClientId,
 			nextClientId: _nextClientId,
 			orientation:
 				getBlockListSettings( insertionPoint.rootClientId )
 					?.orientation || 'vertical',
 			rootClientId: insertionPoint.rootClientId,
+			isNavigationMode: _isNavigationMode(),
+			hasReducedUI: settings.hasReducedUI,
```
Comment (mtias) [inline packages/block-editor/src/components/block-tools/insertion-point.js]: Can we update all the terms to be `distractionFree`? It's going to be harder to read the code if we differ longer term

Comment (draganescu) [reply packages/block-editor/src/components/block-tools/insertion-point.js]: I wasn't sure we'd be dropping the other one. But sure yes! :)

**Comment by draganescu** [comment]
@alexstine in [173bb04](https://github.com/WordPress/gutenberg/pull/41740/commits/173bb04dee62e023ee878004116ba5e65acc9d45) I have tackled the toolbar focusing in two ways for the two shortcuts you provided:

- Shift+Tab, handled this via CSS and `:focus-within` and now when a button in the top toolbar gets focus via tabbing the toolbar shows in distraction free mode as well
- Alt+F10, I don't think I can figure out how to show the top toolbar so I added a new exception in `canFocusHiddenToolbar` where we check if the toolbar can be focused at all (e.g. in code edit mode) and the shortcut is disabled.

**Comment by alexstine** [comment]
@draganescu Still not working for me. The shift+tab or alt+f10 in a paragraph block does not allow the block toolbar to gain focus.

Thanks.

File: packages/block-editor/src/components/block-tools/insertion-point.js (hunk 1 above)
Comment (draganescu) [reply packages/block-editor/src/components/block-tools/insertion-point.js]: Done in [6276ad0](https://github.com/WordPress/gutenberg/pull/41740/commits/6276ad015aff2d498600293dbaf9bf5480d6c9bc)

**Comment by draganescu** [comment]
@alexstine there is no block toolbar in distraction free, it's not even rendered. Instead Shift+Tab focuses the editor toolbar, the equivalent of two Shift+Tab key presses in normal mode.

**Comment by alexstine** [comment]
@draganescu So are users just not supposed to format their content in this mode? That does not make much sense to me. I could see hiding the toolbar to have it appear on hover or something, but not completely.

**Comment by jameskoster** [comment]
> So are users just not supposed to format their content in this mode? That does not make much sense to me. I could see hiding the toolbar to have it appear on hover or something, but not completely.

This is why I maintain that this feature should be framed simply as a "Show / Hide UI" toggle.

Then we don't need to implement a bunch of custom functionality to do things like revealing notifications in "distraction free" mode, which feels kind of inherently contradictory.

**Comment by draganescu** [comment]
@jameskoster I think we found some common ground with what @alexstine says by mentioning that:

- on text selection a formattic toolbar appears
- the formattic KBD shortcuts for basic formatting work
- the markdown expanstion for structure (e.g. list, heading) work

So in distraction free there is basic formatting available to users.

**Comment by alexstine** [comment]
@draganescu Just gave the formatting toolbar a test by selecting some text, it allows me to shift+tab to it, so that is a pass.

One last issue. When toggling the mode from options menu, focus gets lost and it is a truly ugly experience. After I select the option, this is what I log for `document.activeElement;`.
```
<body class="wp-admin wp-core-ui js i…support svg sticky-menu">
```

Can you make focus either remain in the menu or go somewhere more useful for enable/disable distraction free mode?

Thanks.

File: packages/edit-post/src/components/header/style.scss [hunk 2]
```diff
@@ -31,6 +31,55 @@
...
+		& > .edit-post-header__toolbar .edit-post-header-toolbar__list-view-toggle {
+			display: none;
+		}
+
+		& > .edit-post-header__settings > .block-editor-post-preview__dropdown {
+			display: none;
+		}
+
+		& > .edit-post-header__settings > .interface-pinned-items {
+			display: none;
```
Comment (mtias) [inline packages/edit-post/src/components/header/style.scss]: Can we combine these display none?

Comment (draganescu) [reply packages/edit-post/src/components/header/style.scss]: Fixed in [e745879](https://github.com/WordPress/gutenberg/pull/41740/commits/e74587928f829a76db497d29df4308fa2343939a)

File: packages/block-editor/src/components/block-tools/selected-block-popover.js [hunk 3]
```diff
@@ -46,6 +47,7 @@ function SelectedBlockPopover( {
 	clientId,
 	rootClientId,
 	isEmptyDefaultBlock,
+	showContents, // we may need to mount an empty popover because we reuse
```
Comment (scruffian) [inline packages/block-editor/src/components/block-tools/selected-block-popover.js:50]:
```suggestion
	showContents, // We may need to mount an empty popover because we reuse.
```

Comment (scruffian) [reply packages/block-editor/src/components/block-tools/selected-block-popover.js:50]: We reuse what?

File: packages/edit-post/src/components/header/style.scss [hunk 4]
```diff
@@ -31,6 +31,53 @@
...
+		}
+
+		& > .edit-post-header__settings > .block-editor-post-preview__dropdown,
+		& > .edit-post-header__settings > .interface-pinned-items {
+			display: none;
+		}
+
+	}
+
+	// we need ! important because we override inline styles
```
Comment (scruffian) [inline packages/edit-post/src/components/header/style.scss]:
```suggestion
	// We need ! important because we override inline styles
```

File: packages/edit-post/src/components/header/style.scss [hunk 5]
```diff
@@ -31,6 +31,53 @@
...
+	// set by the motion component
```
Comment (scruffian) [inline packages/edit-post/src/components/header/style.scss]:
```suggestion
	// set by the motion component.
```

File: packages/edit-post/src/components/header/style.scss [hunk 6]
```diff
@@ -31,6 +31,53 @@
 	}
 }
 
+.is-distraction-free {
```
Comment (scruffian) [inline packages/edit-post/src/components/header/style.scss]: Is there a reason why this appears at this point in the file? Would it be better for it to be at the end?

File: packages/edit-post/src/components/header/style.scss [hunk 7]
```diff
@@ -31,6 +31,53 @@
...
+	.interface-interface-skeleton__header {
+		border-bottom: none;
+	}
+
+	.edit-post-header {
+		position: absolute;
```
Comment (scruffian) [inline packages/edit-post/src/components/header/style.scss]: Nit: You might want to order these alphabetically 🤷

File: packages/edit-post/src/components/header/writing-menu/index.js [hunk 8]
```diff
@@ -37,11 +76,21 @@ function WritingMenu() {
...
 				messageDeactivated={ __( 'Fullscreen mode deactivated' ) }
 				shortcut={ displayShortcut.secondary( 'f' ) }
 			/>
+			<PreferenceToggleMenuItem
+				scope="core/edit-post"
+				name="distractionFree"
+				toggleHandler={ toggleDistractionFree }
+				label={ __( 'Distraction free' ) }
+				info={ __( 'Write with calmness' ) }
+				messageActivated={ __( 'Distraction free mode activated' ) }
```
Comment (scruffian) [inline packages/edit-post/src/components/header/writing-menu/index.js:90]: These messages only appear for me when I use the keyboard shortcuts, not toggling the button. I'm not sure if that's the expected behaviour.

**Review by scruffian — APPROVED** [review]
Left a few nits, but LGTM.

File: packages/edit-post/src/store/actions.js [hunk 9]
```diff
@@ -194,6 +194,19 @@ export const toggleFeature =
...
+ *
+ * @param {string}  feature Feature name.
+ * @param {boolean} value   Feature value
+ */
+export const setFeature =
+	( feature, value ) =>
+	( { registry } ) =>
+		registry
+			.dispatch( preferencesStore )
+			.set( 'core/edit-post', feature, value );
```
Comment (talldan) [inline packages/edit-post/src/store/actions.js]: It'd be good to avoid merging this as a stable API. I plan to deprecate the function above this (`toggleFeature`) in the future in favour of using the preferences API directly.

It should be relatively easy to replace this with:
```js
const { set: setPreference } = useDispatch( preferencesStore );

 // ...
 
 setPreference( 'core/edit-post', 'fixedToolbar', 'false' );
 ```

File: packages/edit-post/src/components/preferences-modal/index.js [hunk 10]
```diff
@@ -95,6 +111,14 @@ export default function EditPostPreferencesModal() {
 								'Customize options related to the block editor interface and editing flow.'
 							) }
 						>
+							<EnableFeature
+								featureName="distractionFree"
```
Comment (talldan) [inline packages/edit-post/src/components/preferences-modal/index.js:114]: Should users who have 'reducedUI' enabled have 'distractionFree' enabled from the start? If so there should be a migration of the setting to the new name.

It can be added to this function - https://github.com/WordPress/gutenberg/blob/trunk/packages/preferences-persistence/src/migrations/preferences-package-data/index.js#L7

Something like:
```js
export default function convertPreferencesPackageData( data ) {
    data = convertReducedUIToDistractionFree( data );
    data = convertComplementaryAreas( data );
    return data;
}
```

And then `convertReducedUIToDistractionFree` would need to be implemented with unit tests. It shouldn't be too complicated. From my memory, `data` will be in this kind of structure:
```js
{
    'core/edit-post': {
        'reducedUI': true // or false
        // ... other preference values
    },
    // ... other editors
}
```

and would need to be converted to this:
```js
{
    'core/edit-post': {
        'distractionFree': true // or false
        // ... other preference values
    },
    // ... other editors
}
```

Comment (draganescu) [reply packages/edit-post/src/components/preferences-modal/index.js:114]: I think not. The difference is too striking to auto migrate this.

Comment (talldan) [reply packages/edit-post/src/components/preferences-modal/index.js:114]: We should look to remove `reducedUI` from the persisted data at some point later if it's unused. No hurry to do that though (it should probably be addressed around the time distraction free lands in core).

Comment (mtias) [reply packages/edit-post/src/components/preferences-modal/index.js:114]: I'd migrate and avoid an orphaned option, reduced UI always felt half done.

File: packages/edit-post/src/store/actions.js (hunk 9 above)
Comment (draganescu) [reply packages/edit-post/src/store/actions.js]: Removed in 9c416dee2c3af689b9ecb6b28034247df0729525

File: packages/edit-post/src/components/preferences-modal/index.js (hunk 10 above)
Comment (draganescu) [reply packages/edit-post/src/components/preferences-modal/index.js:114]: @mtias the orphaned option is a non issue. Like Dan mentioned it can simply be removed. Migrating is a very surprising UI change landing with no warning. Plus we cannot say that there is a  one to one correspondence between reduced UI and DFM and that 100% of users choosing one would choose the other. I think precisely because reduced UI felt half done it should be removed and  _not_ migrate users :)

Edit: I tested and indeed having "reducedUI" on and then landing in this PR just reverts reduced UI, which is really not that much. But suddenly opening an editor with no chrome would be a bad experience IMO.

File: packages/edit-post/src/components/header/writing-menu/index.js (hunk 8 above)
Comment (draganescu) [reply packages/edit-post/src/components/header/writing-menu/index.js:90]: They're the same text but triggered by different things. In the case of DFM the keybiard shortcut handler triggers the message. I think it's good as it stops you from panicking you did something wrong :). We can revise it.

File: packages/edit-post/src/components/header/style.scss (hunk 6 above)
Comment (draganescu) [reply packages/edit-post/src/components/header/style.scss]: Moved it.

**Comment by draganescu** [comment]
I have merged this but it is not prime time ready yet. The PR:

* silently deprecates the reduce UI preference - it has no effect anymore, if it's set
* introduces a new preference for Distraction Free Mode

It still requires a few updates to be at a level on par with other Gutenberg UX:

- [ ] Add a modal welcome for the feature explaining what happened and offering a revert action - this would also solve the focus loss in a better way.
- [ ] The inserter should be available via the top toolbar (exploring an Alfred like UI)
- [ ] The deprecated reduced UI option should be removed from the storage of users who have it
- [ ] Attempt to reduce the number of items in the edit menu (possibly via a segmented control filtering them)

Worst case scenario this could be reverted to an experiment.

**Comment by alexstine** [comment]
Also, since the menu option is a checkbox, the menu should not close after enabling from the Options menu. It is considered bad UX to have one control in a menu act differently than all the others. I'm not really interested as to why this couldn't get done as long as it gets done in the future.

Thanks.

**Comment by talldan** [comment]
It's a very nice feature, well done on getting it to this stage.

Another thing around the options menu (and also other dropdowns for consistency), It might be worth looking at keeping the top bar visible when focus is in the options menu dropdown, as the way it animates can lead to users clicking on the wrong thing:

https://user-images.githubusercontent.com/677833/194989382-26abc5c0-c046-4988-9bf3-b9fcb539fdb2.mp4

File: packages/block-editor/src/components/block-list/use-in-between-inserter.js [hunk 11]
```diff
@@ -15,7 +15,7 @@ export function useInBetweenInserter() {
 	const openRef = useContext( InsertionPointOpenRef );
 	const isInBetweenInserterDisabled = useSelect(
 		( select ) =>
-			select( blockEditorStore ).getSettings().hasReducedUI ||
+			select( blockEditorStore ).getSettings().isDistractionFree ||
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-list/use-in-between-inserter.js:18]: Should we have kept the same key for backward compatibility? I mean sure the new one is better but I think we may avoid deprecations and headaches by just keeping the old name.

File: packages/edit-post/src/components/header/index.js [hunk 12]
```diff
@@ -23,13 +24,12 @@ import MainDashboardButton from './main-dashboard-button';
 import { store as editPostStore } from '../../store';
 import TemplateTitle from './template-title';
 
-function Header( { setEntitiesSavedStatesCallback } ) {
+function Header( { setEntitiesSavedStatesCallback, isDistractionFree } ) {
```
Comment (youknowriad) [inline packages/edit-post/src/components/header/index.js:27]: Why is this a new prop instead of just using useSelect to retrieve it?

File: packages/block-editor/src/components/block-tools/selected-block-popover.js [hunk 13]
```diff
@@ -195,10 +201,14 @@ function wrapperSelector( select ) {
 				?.__experimentalCaptureToolbars
 	);
 
+	const settings = getSettings();
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-tools/selected-block-popover.js:204]: getSettings called here doesn't cause a re-render if the settings change, it should probably be called inside useSelect instead to avoid bugs.

File: packages/edit-post/src/components/header/more-menu/index.js [hunk 14]
```diff
@@ -36,7 +36,7 @@ const MoreMenu = ( { showIconLabels } ) => {
 							scope="core/edit-post"
 						/>
 					) }
-					<WritingMenu />
+					<WritingMenu onClose={ onClose } />
```
Comment (youknowriad) [inline packages/edit-post/src/components/header/more-menu/index.js:39]: How is this change related to this PR?

File: packages/edit-post/src/components/layout/index.js [hunk 15]
```diff
@@ -115,20 +115,28 @@ function Layout( { styles } ) {
...
+			isDistractionFree:
+				select( editPostStore ).isFeatureActive( 'distractionFree' ),
 			showBlockBreadcrumbs: select( editPostStore ).isFeatureActive(
 				'showBlockBreadcrumbs'
 			),
 			// translators: Default label for the Document in the Block Breadcrumb.
 			documentLabel: postTypeLabel || _x( 'Document', 'noun' ),
 		};
 	}, [] );
+	const [ distractionFree, setDistractionFree ] =
```
Comment (youknowriad) [inline packages/edit-post/src/components/layout/index.js:127]: This is a bit weird, why is this a sate and not something coming directly from the useSelect?

File: packages/preferences/src/components/preference-toggle-menu-item/index.js [hunk 16]
```diff
@@ -20,6 +20,8 @@ export default function PreferenceToggleMenuItem( {
 	messageActivated,
 	messageDeactivated,
 	shortcut,
+	toggleHandler = () => null,
```
Comment (youknowriad) [inline packages/preferences/src/components/preference-toggle-menu-item/index.js:23]: maybe onToggle is a better name here?

File: packages/edit-post/src/components/header/writing-menu/index.js [hunk 17]
```diff
@@ -37,11 +76,21 @@ function WritingMenu() {
...
-				info={ __( 'Work without distraction' ) }
+				info={ __( 'Show and hide admin UI' ) }
 				messageActivated={ __( 'Fullscreen mode activated' ) }
 				messageDeactivated={ __( 'Fullscreen mode deactivated' ) }
 				shortcut={ displayShortcut.secondary( 'f' ) }
 			/>
+			<PreferenceToggleMenuItem
+				scope="core/edit-post"
+				name="distractionFree"
+				toggleHandler={ toggleDistractionFree }
```
Comment (youknowriad) [inline packages/edit-post/src/components/header/writing-menu/index.js:87]: Why do we need this handler at all? I thought the `PreferenceToggleMenuItem` component already toggled the preference automatically?

**Review by youknowriad — COMMENTED** [review]
Nice work here, I left some remarks

File: packages/block-editor/src/components/block-list/use-in-between-inserter.js (hunk 11 above)
Comment (mtias) [reply packages/block-editor/src/components/block-list/use-in-between-inserter.js:18]: Do you expect many people to be using the old one?

Comment (youknowriad) [reply packages/block-editor/src/components/block-list/use-in-between-inserter.js:18]: Probably not much but at least we don't need any migration for preferences or anything (unless we don't do anything about these)

Comment (mtias) [reply packages/block-editor/src/components/block-list/use-in-between-inserter.js:18]: I wouldn't do anything, it's a different enough setting to warrant a recommitment

File: packages/edit-post/src/components/header/more-menu/index.js (hunk 14 above)
Comment (draganescu) [reply packages/edit-post/src/components/header/more-menu/index.js:39]: Thanks for the catch up review! 🙏🏻 This change - which I'll probably undo, idk - is to close the writing menu when entering this mode, because the top toolbar hides and this menu remains orphan in the UI.

File: packages/edit-post/src/components/layout/index.js (hunk 15 above)
Comment (draganescu) [reply packages/edit-post/src/components/layout/index.js:127]: I think it is because we need the layout component to rerender so the motion components kick in.

File: packages/edit-post/src/components/header/writing-menu/index.js (hunk 17 above)
Comment (draganescu) [reply packages/edit-post/src/components/header/writing-menu/index.js:87]: Because we need other stuff to happen (e.g. close sidebars) not only to toggle.

File: packages/edit-post/src/components/layout/index.js (hunk 15 above)
Comment (youknowriad) [reply packages/edit-post/src/components/layout/index.js:127]: Still not sure I understand, did you try just removing the state and see if it works properly?

File: packages/edit-post/src/components/header/writing-menu/index.js (hunk 17 above)
Comment (youknowriad) [reply packages/edit-post/src/components/header/writing-menu/index.js:87]: But why is this different than the other preference toggles above? the sidebars should close when the preference is set anyway.

File: packages/edit-post/src/components/keyboard-shortcuts/index.js [hunk 18]
```diff
@@ -46,6 +64,16 @@ function KeyboardShortcuts() {
 			},
 		} );
 
+		registerShortcut( {
+			name: 'core/edit-post/toggle-distraction-free',
+			category: 'global',
+			description: __( 'Toggle disrtaction free mode.' ),
```
Comment (mcsf) [inline packages/edit-post/src/components/keyboard-shortcuts/index.js:70]: Found this typo in the UI, @draganescu:

```diff
-disrtaction
+distraction
```

Comment (talldan) [reply packages/edit-post/src/components/keyboard-shortcuts/index.js:70]: Noticed that too, I fixed it in https://github.com/WordPress/gutenberg/pull/45186

Comment (draganescu) [reply packages/edit-post/src/components/keyboard-shortcuts/index.js:70]: Thank you both!

File: packages/edit-post/src/components/header/writing-menu/index.js (hunk 17 above)
Comment (draganescu) [reply packages/edit-post/src/components/header/writing-menu/index.js:87]: Yes. I did not alter the rendering of everything distraction free should close so that they render conditionally on the preference. Instead I created this cleanup handler. To have:

- the top toolbar disappear
- the inserter close
- the list view close
- the general sidebar close
- the 1st block focused

... when this preference is toggled on I'd have to pepper with checks for the `isDistractionFree` in all these places (the last one would probably still need a special handler). 

I think the answer to:

> why is this different than the other preference toggles above?

is that distraction free has effects across the UI along the lines of availability not only behavior.

**Comment by draganescu** [comment]
Hi @talldan and @alexstine I'll be looking into solving the problem described by Alex via the idea from Dan, to keep the header toolbar "visible" when the focus is in the writing menu. If I can do that, then I don't need to close the writing menu, hence solve Alex's problem too.

**Comment by draganescu** [comment]
The gist is this:

- we animate the header with a motion component
- the motion component has states defined for hover
- to achieve the three animations we nest motion components (header, site hub and tools on the right)
- to allow motion to kick in I have to render it properly setup - not sure why
- so I render a motion component header when distraction free is on and a normal div when distraction free is off
- this unmouts and mounts the whole header hence closing the more menu

I have tried to solve this but have not had any success so far with convincing motion to kick in properly without rerendering the whole header component. I'll keep trying.

PS I know @alexstine said " I'm not really interested as to why this couldn't get done as long as it gets done in the future" - but this is the future and I can't do it :))) there has to be a way, so far I am thinking of just trying with CSS again but framer motion does the easing so much better.

**Comment by alexstine** [comment]
@draganescu As a hot fix, can you simply add some type of `useEffect` hook to listen for the re-render and then focus? Right now, focus is completely lost. If we can't fix the unmounting right now, at least we can fix focus after the mounting happens.

**Comment by draganescu** [comment]
> Right now, focus is completely lost. 

How come, in the code we manually focus the 1st block in the editor specifically to not lose focus?

Opening the more menu dropdown and focus one specific menu item after the header is remounted seems quite convoluted. There has to be a way to avoid the need to re-mounting.

**Comment by alexstine** [comment]
@draganescu To replicate:
1. Open Firefox.
2. Edit a post or page.
3. Select Options and then Distraction free.
4. Notice how the first block is gains focus.
5. Select Options and then Distraction free again.
6. Notice how the whole editor has focus loss.
```
Log
document.activeElement
Log
 <body class="wp-admin wp-core-ui js i…support svg sticky-menu">
```

It seems this only happens when distraction free is disabled so my mistake if this wasn't clear.

Thanks.

**Comment by talldan** [comment]
@draganescu @alexstine I've put a fix together for the focus loss issue in this PR - https://github.com/WordPress/gutenberg/pull/51627.

It seems to resolve the issue in the majority of cases, though I am still seeing a focus loss on the rare occasion.

**Comment by draganescu** [comment]
I've updated and approved that PR 👏🏻

**Comment by alexstine** [comment]
Awesome, having a look now. 👍
